sos/plugins/kubernetes.py

```diff
@@ -15,40 +15,130 @@
+
+            resources = [
+                'events', 'limitrange', 'pods', 'pvc', 'rc',
+                'resourcequota', 'services'
+            ]
+
+            for res in resources:
+                self.add_cmd_output(
+                    "{} {} {}".format(
+                        kube_bin,
+                        kube_get,
+                        res
+                    )
+                )
+            resources.remove('events')
+            # nodes and pvs are not namespaced, must pull separately
+            self.add_cmd_output([
+                "{} get -o json nodes".format(kube_bin),
+                "{} get -o json pv".format(kube_bin)
+                ]
+            )
+
+            if self.get_option('describe'):
+                resources = resources + ['nodes', 'namespaces']
+                for res in resources:
+                    r = self.get_command_output(
+                        "{} get {}".format(
+                            kube_bin,
+                            res
+                        )
+                    )
+                    if r['status'] == 0:
+                        k_list = [k.split()[0] for k in
+                                  r['output'].splitlines()[1:]
+                                  ]
+                        for k in k_list:
+                            self.add_cmd_output(
+                                "{} describe {} {}".format(
+                                    kube_bin,
+                                    res,
+                                    k
+                                )
+                            )
+
+            if self.get_option('podlogs'):
+                r = self.get_command_output(
+                    "{} get pods".format(kube_bin)
+                )
+                if r['status'] == 0:
+                    pods = [p.split()[0] for p in
+                            r['output'].splitlines()[1:]
+                            ]
+                    for pod in pods:
+                        self.add_cmd_output(
+                            "{} logs {}".format(kube_bin, pod)
+                        )
+
+    def postproc(self):
+        # First, clear sensitive data from the json output collected.
+        # This will mask values when the "name" looks susceptible of
+        # values worth obfuscating, i.e. if the name contains strings
+        # like "pass", "pwd", "key" or "token"
+        env_regexp = r'(?P<var>{\s*"name":\s*[^,]*' \
+                     r'(pass|pwd|key|token|cred)[^,]*,\s*"value":)[^}]*'
+        self.do_cmd_output_sub('kubectl', env_regexp,
+                               r'\g<var> "********"', ignore_case=True)
 
+        # Next, we need to handle the private keys and certs in describe
+        # output that is not hit by the previous iteration.
+        key_regexp = r'-----BEGIN ([A-Z]*).*?-----END ([A-Z].*)-----'
+        self.do_cmd_output_sub('describe', key_regexp,
+                               r'********', dotall=True)
```

Comment:
It's a tricky one - in this case it definitely falls on the side of "we're having to bend the existing function so much it needs a new interface".

> For private keys there are ~30 lines, and I know that 30 lines of nothing but Xs would make me go cross-eyed and
> is why I initially elected to nuke the whole key. 

It's just an example. Pick something that strikes a balance between the one extreme of excising the data completely without a trace versus the other of leaving a bunch of ellipsis or whatever in their place. How may lines of substitution you use can either be controlled by the caller or hardwired into the interface - it really doesn't matter that much so long as the reader can tell that something has been removed.

> I know that 30 lines of nothing but Xs

PageDown? :-)

Dunno. Maybe I have spent too much time staring at bits of vmcores and hexdumps to any longer be considered "normal" in this regard.